### PR TITLE
Move RunPython to a separate migration and transaction

### PR DESCRIPTION
**Pull request by ralphiee22**
### Summary

We separate the schema migration from the data migration for allowing guest access. For devices that have already ran the migrations prior to this change, we check if the corresponding schema migration has been applied (removing `guest_access` field from `facility_dataset`). If it has been applied, we skip the data migration because the field has already been removed, and the logic relies on that field still existing. 

### Reviewer guidance

Any possible problems with the simple approach?

### References

Fixes https://github.com/learningequality/kolibri/issues/6480.

----

### Contributor Checklist

PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by benjaoming** [comment]
I was wondering if having several transactions in one migration is a good idea? Does it already happen? Wouldn't that make the migration none-atomic?

**Comment by ralphiee22** [comment]
After testing the behavior, if `RunPython` fails, it appears that the whole migration is rolled back (the sub-transactions are in a larger transaction?). 
Also interesting https://docs.djangoproject.com/en/1.11/howto/writing-migrations/#non-atomic-migrations

**Comment by benjaoming** [comment]
@ralphiee22 I tried skimming the docs and I found in 2.2 docs that it seems like RunPython will likely still be problematic? I haven't read its definition in source code.



> By default, RunPython will run its contents inside a transaction on databases that do not support DDL transactions (for example, MySQL and Oracle). This should be safe, but may cause a crash if you attempt to use the schema_editor provided on these backends; in this case, pass atomic=False to the RunPython operation.
> 
> On databases that do support DDL transactions (SQLite and PostgreSQL), RunPython operations do not have any transactions automatically added besides the transactions created for each migration. Thus, on PostgreSQL, for example, you should avoid combining schema changes and RunPython operations in the same migration or you may hit errors like OperationalError: cannot ALTER TABLE "mytable" because it has pending trigger events.

https://docs.djangoproject.com/en/2.2/ref/migration-operations/#runpython

**Comment by ralphiee22** [comment]
yeah I glossed over those docs as well, and was a bit confused by what they are saying and what I'm seeing. 
Anyways to mitigate all concerns, it would be best to separate out the schema and data migrations.

**Comment by benjaoming** [comment]
I have had discussions about this previously, and I have this stance: It is okay to edit migrations. You can find other takes on this :)

Now, if we do edit `0005_auto_20191203_0951.py` and remove the `RunPython` part, we would have to add a new conditional data migration - one that takes into account that it may have been run previously (before `0005_auto_20191203_0951.py` was updated)

What would you think about that?

**Comment by ralphiee22** [comment]
Is there any harm in just executing the `RunPython` again in a `0006` migration, regardless if it has run before or not? just trying to understand why we need the conditional check.

**Comment by benjaoming** [comment]
No, the field `allow_guest_access`that is used to check previous settings has been removed here:

https://github.com/learningequality/kolibri/blob/develop/kolibri/core/auth/migrations/0017_remove_facilitydataset_allow_guest_access.py

Here is the definition of the RunPython:

```
def migrate_allow_guest_access(apps, schema_editor):
    FacilityDataset = apps.get_model("kolibriauth", "FacilityDataset")
    DeviceSettings = apps.get_model("device", "DeviceSettings")
    try:
        # Firstly we try to get the default facility's settings
        default_facility = DeviceSettings.objects.get().default_facility
        allow_guest_access = default_facility.dataset.allow_guest_access
    except ObjectDoesNotExist:
        # If it isn't set, we just find out if any facilities had guest access disabled
        # ...if no one disabled it, we will allow it
        allow_guest_access = (
            FacilityDataset.objects.filter(allow_guest_access=False).count() <= 0
        )
    DeviceSettings.objects.update(allow_guest_access=allow_guest_access)
```

If you run this twice, and `allow_guest_access` had been set manually in the meantime, it would get overwritten.

One good thing: For blank installations running all the migrations (or some squashed future version), all these decisions don't matter.

**Comment by ralphiee22** [comment]
ahh right, didn't see that.
> Now, if we do edit 0005_auto_20191203_0951.py and remove the RunPython part, we would have to add a new conditional data migration - one that takes into account that it may have been run previously (before 0005_auto_20191203_0951.py was updated)

This approach sounds good moving forward.

File: kolibri/core/device/migrations/0006_migrate_guest_access.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+from django.db import models
+from django.db.migrations.recorder import MigrationRecorder
+
+
+def migrate_allow_guest_access(apps, schema_editor):
+    # if this migration has been run, drop out because logic depends on a field that no longer exists
+    if MigrationRecorder.Migration.objects.filter(
+        app="kolibriauth",
+        name="0017_remove_facilitydataset_allow_guest_access",
+        applied__isnull=False,
```
Comment (benjaoming) [inline kolibri/core/device/migrations/0006_migrate_guest_access.py:15]: Why `__isnull=False`? Could it be better with a `.exclude(applied=True)` ?

**Review by benjaoming — APPROVED** [review]
Great solution! Would be nice with a confirmation from @mrpau-dev (currently reproducing the issue with Postgres) that it works, so would suggest to keep the PR open until then.

File: kolibri/core/device/migrations/0006_migrate_guest_access.py (hunk 1 above)
Comment (ralphiee22) [reply kolibri/core/device/migrations/0006_migrate_guest_access.py:15]: `applied` is actually a datetime field. It is not null when the migration has been applied.

Comment (benjaoming) [reply kolibri/core/device/migrations/0006_migrate_guest_access.py:15]: Ah of course! Thanks!

**Pull request merged**